File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedAttack.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.Player;
import cn.nukkit.math.Mth;
import cn.nukkit.utils.BinaryStream;
import cn.nukkit.utils.TextFormat;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.playback.PlayerPlaybackEngine;
import net.easecation.ghosty.recording.player.PlayerRecordNode;

public class PlayerUpdatedAttack implements PlayerUpdated {

    private long attackTarget;

    public static PlayerUpdatedAttack of(long attackTarget) {
        return new PlayerUpdatedAttack(attackTarget);
    }

    @Override
    public int getUpdateTypeId() {
        return PlayerUpdated.TYPE_ATTACK;
    }

    @Override
    public boolean hasStates() {
        return false;
    }

    @Override
    public void processTo(PlaybackNPC ghost) {
        PlayerPlaybackEngine engine = ghost.getEngine();
        if (engine != null && engine.displayAttackDistance) {
            // 从世界中找到目标实体
            ghost.getLevel().getActors().values().stream()
                .filter(e -> e instanceof PlaybackNPC)
                .filter(e -> ((PlaybackNPC) e).getOriginEntityId() == attackTarget)
                .findFirst()
                .ifPresent(e -> {
                    // 计算距离
                    double distance = e.distance(ghost);
                    String distanceStr = getDistanceString(distance);
                    // 显示距离（包含双方的Ping数据）
                    String msg = "[Attack] " + distanceStr + TextFormat.WHITE + " "
                        + "[" + PlayerUpdatedPing.getDisplayPing(ghost.lastPing) + TextFormat.WHITE + "]" + ghost.getAliasName()
                        + TextFormat.RESET + TextFormat.WHITE + " -> "
                        + "[" + PlayerUpdatedPing.getDisplayPing(((PlaybackNPC) e).lastPing) + TextFormat.WHITE + "]" + ((PlaybackNPC) e).getAliasName();
                    for (Player viewer : ghost.getViewers().values()) {
                        viewer.sendMessage(msg);
                    }
                });
        }
    }

    public static String getDistanceString(double distance) {
        if (distance >= 3.6) {
            return TextFormat.RED.toString() + Mth.round(distance, 4);
        } else if (distance >= 3) {
            return TextFormat.YELLOW.toString() + Mth.round(distance, 4);
        } else {
            return TextFormat.GREEN.toString() + Mth.round(distance, 4);
        }
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        // 不需要在init时应用
        return node;
    }

    public PlayerUpdatedAttack(BinaryStream stream) {
        read(stream);
    }

    private PlayerUpdatedAttack(long attackTarget) {
        this.attackTarget = attackTarget;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof PlayerUpdatedAttack o)) return false;
        return attackTarget == o.attackTarget;
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putEntityRuntimeId(this.attackTarget);
    }

    @Override
    public void read(BinaryStream stream) {
        this.attackTarget = stream.getEntityRuntimeId();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedAttack{" +
                "attackTarget=" + attackTarget +
                '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedAttack.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.math.Mth
import cn.nukkit.math.Vector3
import cn.nukkit.utils.BinaryStream
import cn.nukkit.utils.TextFormat
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.playback.AttackDistanceCompensationResult

internal fun Double.getDistanceString(): String {
    val color = when {
        this >= 3.6 -> TextFormat.RED
        this >= 3.0 -> TextFormat.YELLOW
        else -> TextFormat.GREEN
    }
    return color.toString() + Mth.round(this, 4)
}

@Serializable
data class PlayerUpdatedAttack(
    val attackTarget: Long,
) : PlayerUpdated {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_ATTACK

    override fun processTo(ghost: PlaybackNPC) {
        val engine = ghost.engine ?: return
        if (!engine.displayAttackDistance) return
        val victim = ghost.level?.entities
            ?.filterIsInstance<PlaybackNPC>()
            ?.firstOrNull { it.originEntityId == attackTarget } ?: return
        // 名字可能包含换行，显示时需要替换为单行
        // ghost是攻击者，victim是受击者
        val attackerName = ghost.nameTag.replace('\n', ' ')
        val victimName = victim.nameTag.replace('\n', ' ')
        val message = buildString {
            val distance = ghost.distance(victim).getDistanceString()
            val compensatedDistance = when (
                val compensation = engine.levelPlaybackEngine?.attackDistanceCompensator?.calculate(
                    engine,
                    attackTarget,
                    engine.tick,
                    Vector3(ghost.x, ghost.y, ghost.z),
                )
            ) {
                is AttackDistanceCompensationResult.Available -> compensation.distance.getDistanceString()
                is AttackDistanceCompensationResult.Unavailable,
                null -> TextFormat.GRAY.toString() + "N/A"
            }
            append("[Attack] ")
            append("标准:").append(distance).append(TextFormat.WHITE).append(" ")
            append("延迟补偿攻击距离:").append(compensatedDistance).append(TextFormat.WHITE).append(" ")
            append("[${PlayerUpdatedPing.getDisplayPing(ghost.lastPing)}${TextFormat.WHITE}]${attackerName}")
            append("${TextFormat.RESET}${TextFormat.WHITE} -> ")
            append("[${PlayerUpdatedPing.getDisplayPing(victim.lastPing)}${TextFormat.WHITE}]${victimName}")
        }
        ghost.viewers.values.forEach { it.sendMessage(message) }
    }

    companion object {
        @JvmStatic
        fun of(attackTarget: Long) = PlayerUpdatedAttack(attackTarget)

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedAttack> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedAttack> {
        override fun write(updated: PlayerUpdatedAttack, stream: BinaryStream) {
            stream.putEntityRuntimeId(updated.attackTarget)
        }

        override fun read(stream: BinaryStream): PlayerUpdatedAttack =
            PlayerUpdatedAttack(stream.entityRuntimeId)
    }
}
